README.md
Description of the scripts in this directory

General
=======

For the scipts in this directory, where it makes sense (and as I update them), I have devised the following common directory path for the output 

/Data  	       	       		<---- the parent of all your collected data
 |---Forensics			<---- parent of forensics activities
     |---Case Identifier	<---- some number or name for the case
     	 |---images		<---- disk images and raw evidence data
	 |---analysis		<---- output of tools and notes
	 |---malware		<---- discovered malicious code samples
	 |---logs		<---- other supporting logs (e.g. network traffic)
 |---Hashes			<---- various hash databases
     |---NSRL			<---- NSRL data files storage
     |---Malware		<---- malware hash files
 |---Wordlists			<---- dictionaries and password wordlists
 |---Malware			<---- malware samples repository

create_image 
============ 

Create image is a shell script to prompt for
some relevant categorical metadata regarding a physical storage
device.  It takes this information to add to an disk inventory and
create a forensic image of the device to be used for further analysis.

Currently, the script is tailored for Windows bitlocker drives only and requires libbde:  https://code.google.com/p/libbde

Your examination system should be configured to NOT automatically mount devices connected to it.


disk_examine
============

Disk examine is a shell script to create an image and perform initial examination on a disk image.  This is intended to be run against a previously captured drive image in raw (dd) format.

Setup
-----
The script makes certain assumptions about the operating system layout upon which it is run.  If you do not want to conform, you can simply change the relevant parts of the script.  Otherwise, you can make changes so you have the following:

 - multiple 'examine' directories under mnt numbered sequentially (e.g. /mnt/examine1, /mnet/examine2...)
 - a /Data directory  (with a symlink from /data for typos)
 - a /Data/hashes directory that contains white and black list hash files, and also acts as a repository for collected hashes for later analysis

Requirements:
-------------
 - bash
 - plaso (log2timeline)
 - libc 2.15+ (on Debian, you need to use testing repo)
   - add 'deb http://ftp.debian.org/debian testing main' to apt sources
   - run 'apt-get -t testing install libc6-amd64 libc6-dev libc6-dbg'
 - md5deep
 - ssdeep
 - disable automount of connected devices

Optional
--------
These are not strictly required, but highly recommended
 - ClamAV
 - AVG (http://free.avg.com/us-en/download-free-all-product)
 - f-prot (installed under /opt/f-prot)


How it works
------------
This script simply automates a number of tasks you would perform anyway in order to examine a disk image or preserve data for later review or outside consultants at such a time you realize that system you ignored became interesting due to new information.

The functions are in alphabetical order.

Enviro-prep
===========
A laughable script that helps to prepare your environment.

NSRL fetch
==========


CHANGELOG.org
* Forensics and eDiscovery scripts Changelog
  This file describes completed and planned changes for the various scripts in this directory
** Version 0.5
*** General
**** Features
***** TODO Case metadata persistence
      Remember some attributes of a case and use them as defaults
***** TODO Configuration File
      Currently, the default locations are hard-coded.  Provide the ability for the end-user to specify these defaults in a specific file
**** Analyze image
***** Bugs
***** Features
****** TODO YARA integration
** Version 0.4
*** Analyze image (analyze_image.sh)
**** Structure 
**** Features
***** Identify OS type
****** TODO Windows 7
****** TODO Windows 8
****** TODO MacOSX
***** Hashing
****** TODO Target specific directories based on OS type
****** TODO (option to) remove known good files (based on hash) from archived storage
****** TODO Review the Hashing structure
****** TODO Add to hash directories automatically
***** Timeline
*** Create image (create_image.sh)
**** Structure
**** Features
*** Copy Evidence (copy_evidence.sh)
**** Bugs
***** TODO Needs to handle situation where a file already exists at destination
*** NSRL (nsrl_fetch.sh)
**** TODO Get the script working to download initial content
*** Readme file
**** TODO Migrate to Emacs Org format
** Version 0.3
   This is really the first "version" where most of the scripts are in working order
*** Analyze image (review_drive.sh)
**** Structure
***** TODO rename from review_drive.sh to analyze_image.sh
***** TODO change destination and working directories to start with capital letters
***** DONE changed from '/bin/sh' to '/bin/bash' for processing
***** DONE remove the image section to a different script
**** Features
   - added log2timeline/plaso
   - added checks for AVG, f-prot, & Clam (current Clam options require 0.98
   - Manifest collects more information about tasks
   - re-worked the directory structure
   - added some supporting documentation created at run-time
*** Create image (create_image.sh)
**** Bugs
***** TODO Fix file-handling to deal with spaces in filenames
**** Structure
***** DONE Add an archive destination
      - This stores a compressed copy of the image in a specified archive location
      - Archives default to 7zip but fall-back to gzip
**** Features
***** DONE Inventory file
      - tracks images created
      - updates file archive location
***** DONE Drive information
      - Physical drive and logical partition info collected
      - information is added to log file
      - Considerably lowered 7z compression level
      - modified many defaults for more streamlined workflow
      - added Checksum field to inventory; Image sum is already stored in log
      - images now archived to images subdirectory


review_drive.sh
#!/bin/bash
###
# This is the InfoSec automated drive examination script v 0.3-beta
# created by Kevin Neely
#
# -= Functionality =-
# Generally, this script performs the following:
# 1) mount the provided disk image at $EXAMINE_DIR, 
# 2) perform some automated analysis, 
# 3) extract potentially relevant data from the image, and
# 4) package everything up in an archive for preservation
#
# -= Version History =-
# 0.3 Change of strategy -- IN PROGRESS
#  - changed from 'sh' to 'bash' for processing
#  - remove the image section to a different script
#  - added log2timeline/plaso
#  - added checks for AVG, f-prot, & Clam (current Clam options require 0.98
#  - Manifest collects more information about tasks
#  - re-worked the directory structure
#  - added some supporting documentation created at run-time
# 0.2 Revisions
#  - added info collection questionnaire
#  - added case name collection
#  - creates an inventory entry in the image repository
#  - added AVG virus scanning
# 0.1 Initial Version
#  - makes image with dc3dd to temp location (if available, else dd)
#  - mounts the image read-only
#  - extracts event logs 
#  - creates hash values of interesting files
#  - scans drive with clamscan
#  - compresses image and logs into a single file on storage location
#
# -= Usage=-
# Usage: review_drive.sh </path/to/block device>
# Example: examine.sh /Data/case1/images/laptop.dd
#
# Requirements:
# To run the script, create the following setup on your system:
# Directory structure as defined in the readme
# Plaso (log2timeline) $git clone https://code.google.com/p/plaso/
# 
###

# ingest the command line
IMAGE=$1


# Functions    

archive () {      
    echo -e "\n \n Compress the image and associated files and copy to archival system"
    echo -e "-----------------------------------------\n"
    7z a $IMAGE_DIR/$SYSTEM -mx=9 -w$TEMP $ANALYSIS_DIR/Manifest.lst $ANALYSIS_DIR/$SYSTEM.log $ANALYSIS_DIR/$SYSTEM.hlog $ANALYSIS_DIR/$IMAGE
    echo 
    }

check_program () {  # checks is a program exists on the system
    echo "check for existence of $1"
    type "$1" &> /dev/null
}

collect_info () {
    echo "Enter the case identifier. (def: [new_case]"
    read input_case
    CASE_NAME=${input_case:=new_case}
    echo "What is the name of the custodian? (def: corpuser)"
    read custodian_name
    CUSTODIAN=${custodian_name:=corpuser}
    echo "What is the system model? (def: laptop)"
    read system_model
    SYSTEM_TYPE=${system_model:=laptop}
    SYSTEM=$CUSTODIAN-$SYSTEM_TYPE
    TEMP=/Data/tmp/$SYSTEM
    echo "What directory should be used for storing the forensics ouput? (def: /Data/Forensics/$CASE_NAME)"
    read root_dir
    FORENSICS_DIR=${root_dir:=/Data/Forensics/$CASE_NAME}
    ANALYSIS_DIR=$FORENSICS_DIR/analysis
    echo "What is the approximate date/time of interest?  YYYY-MM-DD HH:MM:SS"
    read event_time
    }

compare_hash () {
    # compare the hashes against known files.  This needs a re-write and not currently called by the main program
    # NEEDS SOME SERIOUS REWORK
    # compare MD5
    echo "checking for md5deep matches"
    echo -e "MD5sum matches: \n" > $ANALYSIS_DIR/hash_match.log >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    md5deep -r -i 4M -M ~/malware/analysis/malware.md5sums $EXAMINE_DIR/Documents\ and\ Settings/$VICTIM/* $EXAMINE_DIR/WINDOWS/system32/* >> $ANALYSIS_DIR/hash_match.log
    # compare SHA1
    # compare SSDEEP
    # if hash_match <> 0 bytes, then
    echo -e "--===HASH MATCH LOG===-- \n"
    cat $ANALYSIS_DIR/hash_match.log >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
#    7z a $IMAGE_DIR/$SYSTEM -mx=9 $ANALYSIS_DIR/hash_match.log
    }

create_docs () {
    echo -e "============================================\n" > $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    echo -e "InfoSec automated examination script\n" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    echo -e "$SYSTEM \t $CASE_NAME \t $(date +%Y%m%d-%T)" >> $FORENSICS_DIR/../inventory.txt
    echo -e "Disk examination directory structure: \n"  > $FORENSICS_DIR/Readme.txt
    echo -e " |---./			<---- parent dir" >> $FORENSICS_DIR/Readme.txt
    echo -e "     |---$CASE_NAME	<---- case/incident identifier" >> $FORENSICS_DIR/Readme.txt
    echo -e "     	 |---images		<---- disk images and raw evidence data" >> $FORENSICS_DIR/Readme.txt
    echo -e "	 |---analysis		<---- output of tools and notes" >> $FORENSICS_DIR/Readme.txt
    echo -e "	 |---malware		<---- discovered malicious code samples" >> $FORENSICS_DIR/Readme.txt
    echo -e "	 |---logs		<---- other supporting logs (e.g. network traffic)" >> $FORENSICS_DIR/Readme.txt
    }

create_timeline () {
    echo -e "\n \n Examine system with plaso to create a timeline"
    echo -e "-------------------------------------------------\n"
    if check_program log2timeline; then
	log2timeline.py -i $ANALYSIS_DIR/$SYSTEM-timeline.zip $IMAGE
	echo -e "$SYSTEM-timeline.zip\n" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	pinfo.py $ANALYSIS_DIR/$SYSTEM-timeline.zip >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	echo -e "\n \n \n" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    else echo "!!log2timeline not found. Please install plaso and log2timeline!!"
    fi
}

eventLogs_WinXP () {
echo "copy Windows XP event logs"
cp $EXAMINE_DIR/WINDOWS/SYSTEM32/CONFIG/SecEvent.Evt $ANALYSIS_DIR/$SYSTEM-SecEvent.Evt
cp $EXAMINE_DIR/WINDOWS/SYSTEM32/CONFIG/SysEvent.Evt $ANALYSIS_DIR/$SYSTEM-SysEvent.Evt
cp $EXAMINE_DIR/WINDOWS/SYSTEM32/CONFIG/AppEvent.Evt $ANALYSIS_DIR/$SYSTEM-AppEvent.Evt
echo "compress logs"
#7z a $IMAGE_DIR/$SYSTEM -mx=9 $ANALYSIS_DIR/*.Evt
    }

eventlogs_Win7 () {   # should consolodate these into one
    echo "Win7"
}


hash_files () {
# This section needs some work.  
    echo "Record hash values of interesting file system areas" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    if check_program md5deep; then
	echo "running md5deep"
	echo -e "\n create md5 sums"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	echo "---------------" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	md5deep -r -i 4M $EXAMINE_DIR/Windows/System32/* > $ANALYSIS_DIR/$SYSTEM.md5sums
	echo "md5 sums written to $SYSTEM.MD5SUMS"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    else echo "!!md5deep not found, please install!!"
    fi
#   md5deep -r -i 4M  $EXAMINE_DIR/Documents\ and\ Settings/$VICTIM/* $EXAMINE_DIR/WINDOWS/system32/* > $ANALYSIS_DIR/$SYSTEM.md5sums
#    echo "sha1 sums" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
#    sha1deep  -r -i 4M  $EXAMINE_DIR/Documents\ and\ Settings/$VICTIM/* $EXAMINE_DIR/WINDOWS/system32/* > $ANALYSIS_DIR/$SYSTEM.sha1sums
#    echo "ssdeep sums" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
#    ssdeep -r $EXAMINE_DIR/Documents\ and\ Settings/$VICTIM/Local\ Settings/Temporary\ Internet\ Files/ $EXAMINE_DIR/Documents\ and\ Settings/$VICTIM/Local\ Settings/Temp/   $EXAMINE_DIR/WINDOWS/system32/ > $ANALYSIS_DIR/$SYSTEM.ssdeep
#    7z a $IMAGE_DIR/$SYSTEM -mx=9 $ANALYSIS_DIR/$SYSTEM.ssdeep $ANALYSIS_DIR/$SYSTEM.md5sums  $ANALYSIS_DIR/$SYSTEM.sha1sums
#    cp $ANALYSIS_DIR/$SYSTEM.sha1sums /Data/hashes/endpoints
#    cp $ANALYSIS_DIR/$SYSTEM.md5sums /Data/hashes/endpoints
#    cp $ANALYSIS_DIR/$SYSTEM.ssdeep /Data/hashes/endpoints
    }


mount_image () {
# look for empty 'examine' dir under /mnt
echo "Attempting to mount device $1" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
blkid $IMAGE >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
for (( i = 1 ; i <= 8; i++ ))
do
    EXAMINE_DIR="/mnt/examine$i";
    if find $EXAMINE_DIR -maxdepth 0 -empty | read v;  # find an unused mount point
    then 
	echo "empty examination directory found at /mnt/examine$i";
	echo "$IMAGE"
	mount -o ro,loop $IMAGE /mnt/examine$i; # mount the image read-only
	echo "Mounted $IMAGE at /mnt/examine$i" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	i=200;   # set i arbitrarily high to break the loop
    else 
	echo "Mounting $IMAGE at /mnt/examine$i failed" 
    fi
done
    }


scanimage () {
    echo -e "\n \n Scan the mounted image with available AV scanners" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    echo -e "----------------------------------------------------\n" >>  $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    echo -e "Creating $FORENSICS_DIR/malware for storage of infected files\n" >>  $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    mkdir -p $ANALYSIS_DIR/malware
    if check_program clamscan; then
	echo "starting ClamAV scan" >>  $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	clamscan -r --infected --copy=$ANALYSIS_DIR/malware --log=$ANALYSIS_DIR/$SYSTEM-clam.log $EXAMINE_DIR
	cat $ANALYSIS_DIR/$SYSTEM-clam.log >>  $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	echo "clamAV scan complete" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    else echo "clamAV scanner not found"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    fi
    if check_program avgscan; then
	echo "starting AVG scan" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	avgscan -Hac --report=$ANALYSIS_DIR/$SYSTEM-avg.log $EXAMINE_DIR
	echo "AVG scan complete"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    else echo "AVG scanner now found"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    fi
    if check_program fpscan; then
	echo "starting F-prot scan" >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
	/opt/f-prot/fpscan --report --adware --applications -u 3 -s 3 --output=$ANALYSIS_DIR/$SYSTEM-fp.log $EXAMINE_DIR;
	echo "F-prot scan complete"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst;
    else echo "F-prot scanner not found"  >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    fi
    }

timeline_review () {
# TODO
# extract common "interesting" information
# ask the examiner for the date & time of interest 
  # extract events around the time of interest
    echo " "
    psort.py --slice "$event_time" --slice_size 10 -o L2tcsv -w $ANALYSIS_DIR/$SYSTEM-timeline_hits.csv $ANALYSIS_DIR/$SYSTEM-timeline.zip >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
    psort.py -o L2tcsv -w $ANALYSIS_DIR/$SYSTEM-timeline_hits.csv $ANALYSIS_DIR/$SYSTEM-timeline.zip "parser is 'WinJobParser'" |grep -v Google >> $ANALYSIS_DIR/$SYSTEM-Manifest.lst
}

###
# begin script
# 
if [ -z "$1" ]
    then
    echo " "
    echo "Welcome to -insert clever acronym here-"
    echo " "
    echo "Generally, this script performs the following:"
    echo "1) mounts the device specified on the command line"
    echo "2) perform some automated analysis, "
    echo "3) extract potentially relevant data from the image, and"
    echo "4) package everything up in an archive for preservation"
    echo " "
    echo -e "Please provide a disk image for analysis.\n"
    echo -e "Usage: review_drive.sh /path/to/raw_image.dd\n"
    exit
fi
# execute functions
collect_info 
mkdir -p $ANALYSIS_DIR
create_docs # create supporting & descriptive documentation
mount_image
# eventLogs 
# eventLogs_WinXP
hash_files # create hash of temp and sys32 files
# compare_hash # check generated hashes against known bad files
scanimage  # scan the image with AV scanners
create_timeline # generate timeline of events on system
timeline_review 
umount $EXAMINE_DIR
#archive

echo "clean up!" 
# currently, this is just a manual imperative; will eventually clean up extraneous files leftover
